### PR TITLE
Fixed "Add new patient" link 

### LINKED ISSUE #3100: Add new patient not working
### Describe the bug

"Click Here to add a new patient." link not redirecting to the add new patient page.

### To Reproduce
image

1. Hover on  'Administration/Patients'
2. Click on 'Merge Patient'
3. Click on Input feild next to target patient 
4. Search any patient which doesn.t exist.
5. Click on "Click Here to add a new patient." 

image

### Expected behavior
Should redirect current page to "Add new patient" page.

[In another senerio] Currently onclick of this button it redirects the page to the "Add new patient" in Same page. IMO that new page should open in another window or Tab.  Otherwise the current state of tab is getting lost 🤔 

### Client configuration

- Browser: CHrome
- Operating system: Win 10

### Additional Context
I would like to contribute on this 😄

### PR DESCRIPTION
**Pull request by yashrajbothra**
Signed-off-by: Yash Bothra <yashrajbothra786@gmail.com>

Fixes #3100 

#### Short description of what this resolves:
1. Change the warning message 
2. Change the behaviour of "Add new patient" link

### CODE REVIEW

**Comment by yashrajbothra** [comment]
Hey @sjpadgett Sir, I was little confused with your comment https://github.com/openemr/openemr/issues/3100#issuecomment-596108541 so I tried to implement something similar. Please guide if i did something very bad.

**Comment by bradymiller** [comment]
hi @yashrajbothra ,

The problem with supporting entering new patient in modal is some odd things can happen, such as:
image
image

Probably also potential to not have "clean" change over to the new patient (ie. mixing of pids in some screens potentially leading to entering in data for wrong patient).

Seems like easiest thing would be to either:
1. Not support adding new patient from the calendar gui (ie. just do what the merge pt search does)
2. Or go directly to the new patient tab as is typically done in OpenEMR since there is so much code baked into doing it that way. However this would clear the calendar modals thus potentially the scheduler would lose data entered, albeit the scheduler would likely just make that mistake once.

Unless @sjpadgett has input on other options or if feels can robustly support adding a new patient from a modal.

**Comment by yashrajbothra** [comment]
Hey @bradymiller Sir, What if we identify the new "Add/Search" page is a modal and just create pateint without opening another modal (even to confirm) in it and also there should not be any search option in it. I tried to implement it and seems like i can do that if it looks good 😺?

**Comment by im-Amitto** [comment]
IMHO modal should only be used when we need to notify something or we need to display a one-time form. Using it for forms which needs further inputs or need to display some important information will complicate things and clients may not like it.

**Comment by sjpadgett** [comment]
There's a big difference between dialogs and modals. However, agreed, either should be used sparingly. Not so much for desktop but, they become burdensome in mobile/small devices.

While personally I'd have left the current workflow, as is. However, I don't want to discourage students that are willing to take on the more difficult engineering tasks.
There are several solution to resolve this but we can kill two birds if we maybe tried:
- foremost, as Brady pointed out, we don't want to step on current session. Avoid this by not loading new patient.
- add a hidden div to add_edit_event.php to input minimum new patient information. Will need to be creative on how to handle the link in search alert. 
- create a php function to create/validate patient add.
- on return, refresh existing dialog with new patient pid and name etc.

Also, don't forget we have the standard REST api for patient. Might be easier to use it!

**Comment by bradymiller** [comment]
Put some more thought into this. It's going to be really tough to support a new patient insert in the gui outside the current mechanism since there are so many custom options and globals that are used by the clinic in this functionality to check for duplicates, limit and require certain fields (that we the developers may not even be aware of; such as animal species for vets :) ) and more complicated installs that incorporate the zend new patient form that can do all sorts of crazy things like checking/collecting a central database for a patient id).

Not saying impossible, but the first step would likely entail cleaning up and modularizing the "mess" that current exists in the new patient mechanism and gui.

Another option is to actually open a new openemr instance, which would then keep the calendar modals intact (ie. no data entry lost). To see how this is done, check out the 'Open Patient in New Window' toggle (after click the cog) in the patient flow board gui and the 'Open in New Window' toggle in the patient finder gui. In those cases, it is passing the patient to the new instance to open the new patient in the new instance. In your case, would just pass a setting for the new instance to just open the new patient entry form in the first tab (and not open a second tab). Then the scheduler could enter in the new patient in the new isntance and go back to the old instance and then the new patient would show up when they search again.

**Comment by yashrajbothra** [comment]
Your idea is very cool but i just implemented something. Better check that off and if it has some edgecases then I will implement the above one 👍

**Comment by yashrajbothra** [comment]
Hey What if we limit user to just "Old Static Add patient page" it will take less tab size and doesnt ask to confirm. I will add a php function so that it doesnt redirect to show user dashboard.

image

**Comment by sjpadgett** [comment]
We currently have several mechanisms for adding new patients that doesn't try to predict the sundry ways to validate new patients. rest api and ccda import names two.
We can simply add a global to turn off add new from appointments. Also, implementation of stricter validation based on setup globals is not that big a deal to cope with.

This PR is a perfect opportunity to add/enhance to our patient controller/service by adding to it's validation mechanism and search/add. This in turn will make cleaning up/changing current views easier along with securing patient adds from rest api and ccda import.

@yashrajbothra 
- explore src/Services/PatientService.php, src/RestControllers/PatientRestController.php and src/RestControllers/RestControllerHelper.php.
- We have easy capability to do internal rest api calls. Check out tests/api/InternalApiTest.php for an example.
- Remember we want to limit the number of open dialogs so recommend adding your Add Patient Record form to a hidden div in add_edit_event.php. Or create a template to add dynamically. Or you could even create using react,twig and on and on.
- Close the search dialog and return to add_edit_event where you'd show/set focus to the hidden div. 
- On a button click to submit new patient then handle with internal rest calls to patient services.
- Of course add any functionality to support to out existing services. Also, we have other services available that may be helpful.

The bottom line is to have a clean patient service workflow that can benefit future modernisation of our views and external api support.

If this is proving a bit overwhelming, I get it. However, if you are willing, so am I. It might be best to do this in stages over a couple PR's so we don't hold up master to long.

**Comment by yashrajbothra** [comment]
Thank you @sjpadgett Sir, I would definetly explore the codebase and will try by best to resolve this issue😀. With this I just wanna ask that can i also refactor some files in the codebase. Do i have to open a issue or direct PR is fine or Can i include in this PR. Whatever suits best will be my first choice 👍

**Comment by sjpadgett** [comment]
> Do i have to open a issue or direct PR is fine

@yashrajbothra 
Really depends on what and how many files. If it is going to take some head scratching to review new PR if minor changes then this PR.

**Comment by yashrajbothra** [comment]
> * Remember we want to limit the number of open dialogs so recommend adding your Add Patient Record form to a hidden div in add_edit_event.php. Or create a template to add dynamically. Or you could even create using react,twig and on and on.

Hey @sjpadgett What fields do i add in **Add Patient Record form** ?

or Should i use https://github.com/openemr/openemr/blob/master/interface/new/new.php (will have to refactor before using) as templete with a check which will not include `new_comprehensive.php` 🤔

**Comment by yashrajbothra** [comment]
Hii @bradymiller Sir , Should i refactor [interface/new/new.php](https://github.com/openemr/openemr/blob/master/interface/new/new.php) this file because code is too old and has some deprecated code and can i include bootstrap in that file? So i can futher use this file in this PR

**Comment by bradymiller** [comment]
hi @yashrajbothra ,
Definitely feel free to ugprade to bootstrap there. Note this script is used when set following to use "Old-style static form without search or duplication check" in Administration->Globals->Appearance (https://www.open-emr.org/wiki/index.php/Administration_Globals#New_Patient_Form).
